### ISSUE TITLE
Requests that go to a different pod are slow

### ISSUE THREAD

**Issue by brettjenkins** [metallb/metallb#649]
So I'm new to MetalLB so it's possible I've done something stupid, but I have this issue which seems extremely odd to me, and smells like a bug.

OS: Freshly installed Ubuntu 18.04 LS

Version v0.9.3

Fresh installation of k8s via kubeadm - 3 node cluster, 1 master, 2 nodes.
BCP mode connected to pfsense using openbgpd

Flannel: v0.12.0-amd64

K8s version: v.1.18.4

Kube-Proxy has been set to ipvs mode with strict on

I've also switched off offloading as this seemed like a similar ish issue https://github.com/kubernetes/kubernetes/issues/88986 but didn't solve my issue

I have deployed ngnix-hello with a replica of 3, which is a simple app which shows the current server and allows constant refreshing - but I have also checked with latest ngnix and had the same issue.

As can be seen from fiddler, most requests are fast. But every 20 or so requests it can take 20-30 seconds to respond. Usually when it has responded the page is then reporting it's coming from a different pod. Accessing the pods via NodePort works fine, not seeing this slowdown once, even when the request gets routed to a different pod from the previous request.

![image](https://user-images.githubusercontent.com/1395855/85483242-5d2e0e80-b5bc-11ea-80a0-bcc11199b94c.png)

Is this a bug?

**Comment by champtar** [comment]
Hi @brettjenkins 

MetalLB is control plane, dataplane is kube-proxy + CNI (flannel), this is a bug in one of those or Linux

Best
Etienne

**Comment by brettjenkins** [comment]
I'm a bit confused though - if I use the Cluster IP from within the cluster, I get no slowdown at all, it's only when going through the LoadBalancer IP I get this issue.

**Comment by champtar** [comment]
Sorry, read too fast, look at the logs, look at the readiness of your service

**Comment by brettjenkins** [comment]
Which logs should I look at?

**Comment by johananl** [comment]
I think @champtar's point was that MetalLB isn't involved in the packet flow, which is true: MetalLB advertises a service over BGP, which makes the upstream router (openbgpd in your case) route that service's external IP to the k8s node on which the MetalLB speaker is running. When the upstream router forwards a packet to the node, the elements which actively process the packet are kube-proxy and the network plugin (Flannel in your case).

Following the above, if the BGP session is stable, it is very unlikely that MetalLB is causing the delays you're experiencing.

To help you with the troubleshooting, I suggest you check the logs of the MetalLB speaker pods and search for any indication that the BGP session(s) are unstable. If they are stable, the problem likely lies elsewhere. If I were you I would run `tcpdump` on one of the nodes which accept the traffic from the router and try to see what's happening with the slow requests: how long does it take to get a SYN/ACK from nginx? Are there any TCP retransmits? Is the traffic crossing Flannel from the ingress node to another node which runs nginx?

Lastly, to get a more interactive help with troubleshooting I suggest you look at the project's [Slack channel](https://metallb.universe.tf/community/#slack).

**Comment by brettjenkins** [comment]
I appreciate your help @johananl - I have tried the slack channel but not had much help as yet

Something weird I have discovered.

All computers that exhibit this hanging request behaviour are Windows machines. Linux machines do not have this issue.

Also, CURL running on a windows machine never has this problem, yet Invoke-WebRequest in PS on the same machine DOES have the problem. How odd!

Also windows VMs hosted on the same server as a couple of the nodes also never have this problem.

I viewed the logs of the NGINX pods, and when the hanging happens, the NGINX pod isn't getting the request, so it seems there's an issue with traffic entering the cluster.

I'm completely confused now, is this a BGP issue? I'm using pfsense with OpenBGPD with the following setup:

`# This file was created by the package manager. Do not edit!

AS 64512
fib-update yes
listen on 10.0.1.18
router-id 10.0.1.18
network 10.26.0.0/22

neighbor 10.0.1.41 { 
	remote-as 64513
        announce all
	descr "gibson-k8s-1" 
}

neighbor 10.0.1.42 { 
	remote-as 64513
        announce all
	descr "gibson-k8s-2" 
}

`

10.0.1.40 is my master, 10.0.1.41 and 10.0.1.42 are my worker nodes in a VM on different physical hosts. 10.0.1.18 is a normal Virtual IP address on pfsense, not sure why this is needed though.

Apologies for being annoying, but I don't know where else to turn!

**Comment by brettjenkins** [comment]
For anyone following this, I think I finally got to the bottom of it. Documenting here, because I can't be the only one to see this.

So I was looking at the pfsense logs and noticed that when I was getting the slow loading it was blocking traffic from the pc making the request to the load balanced ip. I looked into this and eventually found a suggestion to change the firewall optimisation option in the advanced screen. It was set to normal, but I changed it to conservative which "Tries to avoid dropping any legitimate idle connections at the expense of increased memory usage and CPU utilisation", and since doing that it seems to have fixed the problem!

So it seems that pfsense was dropping the connection thinking it was dead when it wasn't, this only seemed to affect windows (and android) but my Linux vms and raspberri pi worked fine. Presumably they handle their connection being dropped better, or pfsense wasn't dropping their connections for some reason.

Either way I've solved it finally! If one good thing has come from this, I've been dropped into the deep end with my k8s education and learnt a lot about how things hang together, so that's good :)

I'll close this ticket now. Thanks for all your help and I really hope this helps someone else (as normal is the default, and I can't be the only person using pfsense with metallb)
